flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeGeneralization.java

```diff
@@ -0,0 +1,631 @@
+				return findCommonMultisetType(normalizedTypes);
+			} else if (typeRoot == ROW) {
+				return findCommonRowType(normalizedTypes);
+			}
+
+			// exit if two completely different types are compared (e.g. ROW and INT)
+			// this simplifies the following lines as we compare same interval families for example
+			if (!areSimilarTypes(resultType, type)) {
+				return null;
+			}
+
+			// for types of family CHARACTER_STRING or BINARY_STRING
+			if (hasFamily(type, CHARACTER_STRING) || hasFamily(type, BINARY_STRING)) {
+				final int length = combineLength(resultType, type);
+
+				if (hasRoot(resultType, VARCHAR) || hasRoot(resultType, VARBINARY)) {
+					// for variable length type we are done here
+					resultType = createStringType(resultType.getTypeRoot(), length);
+				} else {
+					// for mixed fixed/variable or fixed/fixed lengths
+					resultType = createStringType(typeRoot, length);
+				}
+			}
+			// for EXACT_NUMERIC types
+			else if (hasFamily(type, EXACT_NUMERIC)) {
+				if (hasFamily(resultType, EXACT_NUMERIC)) {
+					resultType = createCommonExactNumericType(resultType, type);
+				} else if (hasFamily(resultType, APPROXIMATE_NUMERIC)) {
+					// the result is already approximate
+					if (typeRoot == DECIMAL) {
+						// in case of DECIMAL we enforce DOUBLE
+						resultType = new DoubleType();
+					}
+				} else {
+					return null;
+				}
+			}
+			// for APPROXIMATE_NUMERIC types
+			else if (hasFamily(type, APPROXIMATE_NUMERIC)) {
+				if (hasFamily(type, APPROXIMATE_NUMERIC)) {
+					resultType = createCommonApproximateNumericType(resultType, type);
+				} else if (hasFamily(resultType, EXACT_NUMERIC)) {
+					// the result was exact so far
+					if (typeRoot == DECIMAL) {
+						// in case of DECIMAL we enforce DOUBLE
+						resultType = new DoubleType();
+					} else {
+						// enforce an approximate result
+						resultType = type;
+					}
+				} else {
+					return null;
+				}
+			}
+			// for TIME
+			else if (hasFamily(type, TIME)) {
+				if (hasFamily(resultType, TIME)) {
+					resultType = new TimeType(combinePrecision(resultType, type));
+				} else {
+					return null;
+				}
+			}
+			// for TIMESTAMP
+			else if (hasFamily(type, TIMESTAMP)) {
+				if (hasFamily(resultType, TIMESTAMP)) {
+					resultType = createCommonTimestampType(resultType, type);
+				} else {
+					return null;
+				}
+			}
+			// for day-time intervals
+			else if (typeRoot == INTERVAL_DAY_TIME) {
+				resultType = createCommonDayTimeIntervalType(
+					(DayTimeIntervalType) resultType,
+					(DayTimeIntervalType) type);
+			}
+			// for year-month intervals
+			else if (typeRoot == INTERVAL_YEAR_MONTH) {
+				resultType = createCommonYearMonthIntervalType(
+					(YearMonthIntervalType) resultType,
+					(YearMonthIntervalType) type);
+			}
+			// other types are handled by findCommonCastableType
+			else {
+				return null;
+			}
+		}
+
+		// NULL type only
+		if (resultType == null && hasNullType) {
+			return new NullType();
```

Comment:
I think this branch is unreachable.
`resultType` may be null only if `normalizedTypes` is empty (because of line 234), but then `hasNullType` cannot be `true` as we do not have types at all.